File: indicator.js
class Indicator {
    /**
     * 计算净值指标的类
     *
     * 对外暴露的接口有两个, 一个是初始化方法, 一个是update方法。
     * 初始化方法。如果传入的trade_date和net是数组, 则在实例化时会自动计算净值指标。
     * update方法用于更新净值指标。
     */

    constructor(tradeDate, net, rf) {
        /**
         * 初始化方法。如果传入的trade_date和net是数组, 则在实例化时会自动计算净值指标。
         * 对外暴露的指标属性为：
         *
         * :return_acc: 累计收益率
         * :annual_return_acc: 年化累计收益率
         * :drawdown: 最大回撤
         * :drawdown_start_date: 最大回撤开始日期
         * :drawdown_end_date: 最大回撤结束日期
         * :drawdown_recovery_date: 最大回撤恢复日期
         * :sharp_ratio: 夏普比率
         * :sortino_ratio: 索提诺比率
         * :calmar_ratio: 卡玛比率
         * :return_mean: 日均收益率均值
         * :return_std: 日均收益率标准差
         * :_annual_return_std: 年化收益率标准差
         * :_annual_return_down_std: 年化下行收益率标准差
         */
        this.returnAcc = 0.0;
        this._returnPct = 0.0;
        this._returnSum = 0.0;
        this._returnSquare = 0.0;
        this.returnMean = 0.0;
        this.returnStd = 0.0;

        this.annualReturnAcc = 0.0;
        this.annualReturnPct = 0.0;
        this._sumReturn = 0.0;
        this._downReturnSum = 0.0;
        this._excessSumReturn = 0.0;
        this._num = 0;
        this._downNum = 0;
        this.excessReturnAvg = 0.0;
        this._annualReturnSquare = 0.0;
        this._annualReturnDownSquare = 0.0;
        this._annualReturnStd = 0.0;
        this._annualReturnDownStd = 0.0;

        this.drawdown = 0.0;
        this.drawdownHighSpot = 0.0;
        this._newDrawdown = 0.0;
        this.sharpRatio = 0.0;
        this.sortinoRatio = 0.0;
        this.calmarRatio = 0.0;

        this.duration = 0;

        if (typeof tradeDate === 'string' && typeof net === 'number' && typeof rf === 'number') {
            this._tradeDate = tradeDate;
            this._initDate = new Date(tradeDate);
            this.initNet = net;
            this._net = net;
            this._maxNet = net;
            this._minNet = net;
            this.drawdownStartDate = tradeDate;
            this._lastMaxDate = tradeDate;
            this._lastMinDate = tradeDate;
            this.drawdownEndDate = tradeDate;
            this.drawdownRecoveryDate = tradeDate;
            this.rf = rf;
        } else if (Array.isArray(tradeDate) && Array.isArray(net)) {
            if (tradeDate.length !== net.length) {
                throw new Error("The length of tradeDate and net should be equal");
            }
            this._tradeDate = tradeDate[0];
            this._initDate = new Date(tradeDate[0]);
            this.initNet = net[0];
            this._net = net[0];
            this._maxNet = net[0];
            this._minNet = net[0];
            this.drawdownStartDate = tradeDate[0];
            this._lastMaxDate = tradeDate[0];
            this._lastMinDate = tradeDate[0];
            this.drawdownEndDate = tradeDate[0];
            this.drawdownRecoveryDate = tradeDate[0];
            if (typeof rf === 'number') {
                this.rf = rf;
                for (let i = 1; i < tradeDate.length; i++) {
                    this.update(tradeDate[i], net[i], rf);
                }
            } else if (Array.isArray(rf)) {
                if (rf.length !== tradeDate.length) {
                    throw new Error("The length of tradeDate and rf should be equal");
                }
                for (let i = 1; i < tradeDate.length; i++) {
                    this.update(tradeDate[i], net[i], rf[i]);
                }
            } else {
                throw new Error("Invalid rf type");
            }
        } else {
            throw new Error("Invalid input type");
        }
    }

    get tradeDate() {
        return this._tradeDate;
    }

    get net() {
        return this._net;
    }

    set tradeDate(value) {
        this.duration = this._calDuration(value);
        this._tradeDate = value;
    }

    set net(value) {
        // 计算日均收益均值
        this._num += 1;
        this.returnAcc = value / this.initNet - 1;
        this._returnPct = value / this._net - 1;
        this._returnSum += this._returnPct;
        this._returnSquare += this._returnPct ** 2;
        this.returnMean = this._returnSum / this._num;

        // 计算年化收益率
        this.annualReturnAcc = this._calAnnualReturn(this.returnAcc, this.duration);
        this.annualReturnPct = this._calAnnualReturn(this._returnPct, 1);

        // 计算超额收益率
        this._excessSumReturn += this.annualReturnPct - this.rf;
        this.excessReturnAvg = this._excessSumReturn / this._num;

        // 计算年化收益率、下行收益率和日均受益率的标准差
        this._annualReturnSquare += (this.annualReturnPct - this.rf) ** 2;
        if (this.annualReturnPct < this.rf) {
            this._annualReturnDownSquare += (this.annualReturnPct - this.rf) ** 2;
            this._downReturnSum += this.annualReturnPct - this.rf;
            this._downNum += 1;
        }
        if (this._num > 1) {
            this.returnStd = Math.sqrt((this._returnSquare - this._num * this.returnMean ** 2) / (this._num - 1));
            this._annualReturnStd = Math.sqrt((this._annualReturnSquare - this._num * this.excessReturnAvg ** 2) / (this._num - 1));
            this._annualReturnDownStd = Math.sqrt((this._annualReturnDownSquare - this._downReturnSum ** 2 / this._downNum) / (this._downNum - 1));
        }

        this._net = value;
    }

    _calDuration(tradeDate) {
        /** 计算当前日期距离初始日期的天数 */
        const today = new Date(tradeDate);
        return (today - this._initDate) / (1000 * 60 * 60 * 24);
    }

    _calAnnualReturn(ret, duration) {
        /** 计算年化收益率 */
        return ret * 252 / duration;
    }

    update(tradeDate, net, rf = null) {
        /**
         * 更新净值指标
         *
         * @param {string} tradeDate 交易日期
         * @param {number} net 净值
         * @param {number} [rf] 无风险利率
         */
        if (rf !== null) {
            this.rf = rf;
        }
        this.tradeDate = tradeDate;
        this.net = net;
        this.calDrawdown();
        this.calSharpRatio();
        this.calCalmarRatio();
        this.calSortinoRatio();
    }

    calDrawdown() {
        /**
         * 计算最大回撤
         * 当有多个最低点时取最早的
         * 当有多个最大回撤时取最近的
         */
        if (this.net >= this._maxNet) {
            this._maxNet = this.net;
            this._minNet = this.net;
            this._lastMaxDate = this.tradeDate;
        } else if (this.net < this._minNet) {
            this._minNet = this.net;
            this._lastMinDate = this.tradeDate;
            this._newDrawdown = (this._maxNet - this._minNet) / this._maxNet;
            if (this._newDrawdown > this.drawdown) {
                this.drawdown = this._newDrawdown;
                this.drawdownHighSpot = this._maxNet;
                this.drawdownStartDate = this._lastMaxDate;
                this.drawdownEndDate = this._lastMinDate;
                this.drawdownRecoveryDate = "";
            }
        }
        if (this.net === this.drawdownHighSpot) {
            this.drawdownRecoveryDate = this.tradeDate;
        }
    }

    calSharpRatio() {
        if (this._annualReturnSquare < 0) {
            throw new Error("Return square should be non-negative");
        }
        if (this._num > 1) {
            this.sharpRatio = this.excessReturnAvg / this._annualReturnStd;
        }
    }

    calCalmarRatio() {
        if (this.drawdown > 0) {
            this.calmarRatio = this.annualReturnAcc / this.drawdown;
        }
    }

    calSortinoRatio() {
        if (this._annualReturnSquare < 0) {
            throw new Error("Return square should be non-negative");
        }
        if (this._num > 0 && this._annualReturnDownSquare > 0) {
            this.sortinoRatio = this.excessReturnAvg / Math.sqrt(this._annualReturnDownSquare / this._downNum);
        } else {
            this.sortinoRatio = 0.0;
        }
    }
}

// Example usage:
// const indicator = new Indicator("2023-01-01", 1000, 0.02);
// indicator.update("2023-01-02", 1010);
// console.log(indicator.sharpRatio, indicator.sortinoRatio, indicator.calmarRatio);

File: indicator.py
from typing import Union, Iterable
from datetime import datetime


class Indicator:
    """
    计算净值指标的类

    对外暴露的接口有两个, 一个是初始化方法, 一个是update方法。
    初始化方法。如果传入的trade_date和net是数组, 则在实例化时会自动计算净值指标。
    update方法用于更新净值指标。
    """

    def __init__(
        self,
        trade_date: Union[str, Iterable],
        net: Union[float, Iterable],
        rf: Union[float, Iterable],
    ):
        """
        初始化方法。如果传入的trade_date和net是数组, 则在实例化时会自动计算净值指标。
        对外暴露的指标属性为：

        :return_acc: 累计收益率
        :annual_return_acc: 年化累计收益率
        :drawdown: 最大回撤
        :drawdown_start_date: 最大回撤开始日期
        :drawdown_end_date: 最大回撤结束日期
        :drawdown_recovery_date: 最大回撤恢复日期
        :sharp_ratio: 夏普比率
        :sortino_ratio: 索提诺比率
        :calmar_ratio: 卡玛比率
        ::return_mean: 日均收益率均值
        ::return_std: 日均收益率标准差
        :_annual_return_std: 年化收益率标准差
        :_annual_return_down_std: 年化下行收益率标准差
        """
        self.return_acc: float = 0.0
        self._return_pct: float = 0.0
        self._return_sum: float = 0.0
        self._return_square: float = 0.0
        self.return_mean: float = 0.0
        self.return_std: float = 0.0

        self.annual_return_acc: float = 0.0
        self.annual_return_pct: float = 0.0
        self._sum_return: float = 0.0
        self._down_return_sum: float = 0.0
        self._excess_sum_return: float = 0.0
        self._num: int = 0
        self._down_num: int = 0
        self.excess_return_avg: float = 0.0
        self._annual_return_square: float = 0.0
        self._annual_return_down_square: float = 0.0
        self._annual_return_std: float = 0.0
        self._annual_return_down_std: float = 0.0

        self.drawdown: float = 0.0
        self.drawdown_high_spot: float = 0.0
        self._new_drawdown: float = 0.0
        self.sharp_ratio: float = 0.0
        self.sortino_ratio: float = 0.0
        self.calmar_ratio: float = 0.0

        self.duration: int = 0
        if (
            isinstance(trade_date, str)
            and isinstance(net, float)
            and isinstance(rf, float)
        ):
            self._trade_date = trade_date
            self._init_date = datetime.strptime(trade_date, "%Y-%m-%d")
            self.init_net = net
            self._net = net
            self._max_net = net
            self._min_net = net
            self.drawdown_start_date = trade_date
            self._last_max_date = trade_date
            self._last_min_date = trade_date
            self.drawdown_end_date = trade_date
            self.drawdown_recovery_date = trade_date
            self.rf = rf
        elif isinstance(trade_date, Iterable) and isinstance(net, Iterable):
            assert len(trade_date) == len(
                net
            ), "The length of trade_date and net should be equal"
            self._trade_date = trade_date[0]
            self._init_date = datetime.strptime(trade_date[0], "%Y-%m-%d")
            self.init_net = net[0]
            self._net = net[0]
            self._max_net = net[0]
            self._min_net = net[0]
            self.drawdown_start_date = trade_date[0]
            self._last_max_date = trade_date[0]
            self._last_min_date = trade_date[0]
            self.drawdown_end_date = trade_date[0]
            self.drawdown_recovery_date = trade_date[0]
            if isinstance(rf, float):
                self.rf = rf
                for i in range(1, len(trade_date)):
                    self.update(trade_date[i], net[i], rf)
            elif isinstance(rf, Iterable):
                assert len(rf) == len(
                    trade_date
                ), "The length of trade_date and rf should be equal"
                for i in range(1, len(trade_date)):
                    self.update(trade_date[i], net[i], rf[i])
            else:
                raise ValueError("Invalid rf type")
        else:
            raise ValueError("Invalid input type")

    @property
    def trade_date(self):
        return self._trade_date

    @property
    def net(self):
        return self._net

    @trade_date.setter
    def trade_date(self, value):
        self.duration = self._cal_duration(value)
        self._trade_date = value

    @net.setter
    def net(self, value):
        # 计算日均收益均值
        self._num += 1
        self.return_acc = value / self.init_net - 1
        self._return_pct = value / self._net - 1
        self._return_sum += self._return_pct
        self._return_square += self._return_pct**2
        self.return_mean = self._return_sum / self._num

        # 计算年化收益率
        self.annual_return_acc = self._cal_annual_return(self.return_acc, self.duration)
        self.annual_return_pct = self._cal_annual_return(self._return_pct, 1)

        # 计算超额收益率
        self._excess_sum_return += self.annual_return_pct - self.rf
        self.excess_return_avg = self._excess_sum_return / self._num

        # 计算年化收益率、下行收益率和日均受益率的标准差
        self._annual_return_square += (self.annual_return_pct - self.rf) ** 2
        if self.annual_return_pct < self.rf:
            self._annual_return_down_square += (self.annual_return_pct - self.rf) ** 2
            self._down_return_sum += self.annual_return_pct - self.rf
            self._down_num += 1
        if self._num > 1:
            self.return_std = (
                (self._return_square - self._num * self.return_mean**2)
                / (self._num - 1)
            ) ** 0.5
            self._annual_return_std = (
                (self._annual_return_square - self._num * self.excess_return_avg**2)
                / (self._num - 1)
            ) ** 0.5
            self._annual_return_down_std = (
                (
                    self._annual_return_down_square
                    - self._down_return_sum**2 / self._down_num
                )
                / (self._down_num - 1)
            ) ** 0.5

        self._net = value

    def _cal_duration(self, trade_date: str) -> int:
        """计算当前日期距离初始日期的天数"""
        today = datetime.strptime(trade_date, "%Y-%m-%d")
        return (today - self._init_date).days

    def _cal_annual_return(self, ret: float, duration: int) -> float:
        """计算年化收益率"""
        return ret * 252 / duration

    def update(self, trade_date: str, net: float, rf: float | None = None):
        """
        更新净值指标

        :param trade_date: 交易日期
        :param net: 净值
        :param rf: 无风险利率
        """
        if rf is not None:
            self.rf = rf
        self.trade_date = trade_date
        self.net = net
        self.cal_drawdown()
        self.cal_sharp_ratio()
        self.cal_calmar_ratio()
        self.cal_sortino_ratio()

    def cal_drawdown(self):
        """
        计算最大回撤
        当有多个最低点时取最早的
        当有多个最大回撤时取最近的
        """
        if self.net >= self._max_net:
            self._max_net = self.net
            self._min_net = self.net
            self._last_max_date = self.trade_date

        elif self.net < self._min_net:
            self._min_net = self.net
            self._last_min_date = self.trade_date
            self._new_drawdown = (self._max_net - self._min_net) / self._max_net
            if self._new_drawdown > self.drawdown:
                self.drawdown = self._new_drawdown
                self.drawdown_high_spot = self._max_net
                self.drawdown_start_date = self._last_max_date
                self.drawdown_end_date = self._last_min_date
                self.drawdown_recovery_date = ""
        if self.net == self.drawdown_high_spot and self.drawdown_recovery_date == "":
            self.drawdown_recovery_date = self.trade_date

    def cal_sharp_ratio(self):
        assert self._annual_return_square >= 0, "Return square should be non-negative"
        if self._num > 1:
            self.sharp_ratio = (self.excess_return_avg) / self._annual_return_std

    def cal_calmar_ratio(self):
        if self.drawdown > 0:
            self.calmar_ratio = self.annual_return_acc / self.drawdown

    def cal_sortino_ratio(self):
        assert self._annual_return_square >= 0, "Return square should be non-negative"
        if self._num > 0 and self._annual_return_down_square > 0:
            self.sortino_ratio = (self.excess_return_avg) / (
                self._annual_return_down_square / (self._down_num)
            ) ** 0.5
        else:
            self.sortino_ratio = 0.0
